controller/Publish.php

```diff
@@ -62,4 +71,56 @@ public function wizard()
             $this->setView('DeliveryMgmt/wizard_error.tpl');
         }
     }
+
+    public function selectRemoteEnvironments()
+    {
+        $selectedDelivery = new core_kernel_classes_Resource(
+            tao_helpers_Uri::decode($this->getRequestParameter('uri'))
+        );
+
+        $environments = $this->getServiceLocator()
+            ->get(PublishingService::SERVICE_ID)
+            ->getEnvironments();
+
+        $submitUrl = $this->getServiceLocator()
+            ->get(UrlHelper::class)
+            ->buildUrl('publishToRemoteEnvironment', 'Publish', 'taoPublishing');
+
+        $this->setData('submit-url', $submitUrl);
+        $this->setData('delivery-uri', $selectedDelivery->getUri());
+        $this->setData('delivery-label', $selectedDelivery->getLabel());
+        $this->setData('remote-environments', array_values($environments));
+        $this->setView('PublishToRemote/index.tpl');
+    }
+
+    /**
+     * @param ServerRequest $request
+     * @return mixed
+     * @throws PublishingFailedException
+     */
+    public function publishToRemoteEnvironment(ServerRequest $request)
+    {
+        $requestData = $request->getParsedBody();
+        $deliveryUri = $requestData['delivery-uri'] ?? '';
+        $environments = $requestData['remote-environments'] ?? [];
```

Comment:
I would create a constants for these request parameters that are expected

Reply:
Done.